src/python_testing/TestSpecParsingDeviceType.py

```diff
@@ -0,0 +1,107 @@
+                            '<revision revision="{{ i }}" summary="Rev"/>\n'
+                            '{% endfor %}'
+                            '</revisionHistory>\n'
+                            '<classification {% if classification_class %} class="{{ classification_class }}" {% endif %} {% if classification_scope %} scope="{{ classification_scope }}" {% endif %}/>\n'
+                            '<conditions/>\n'
+                            '<clusters>\n'
+                            '{% for k,v in clusters.items() %}'
+                            '<cluster id="{{ k }}" name="{{ v }}" side="server">\n'
+                            '<mandatoryConform/>\n'
+                            '</cluster>\n'
+                            '{% endfor %}'
+                            '</clusters>\n'
+                            '</deviceType>\n')
+
+        xml = template.render(device_type_id=device_type_id, revision=revision, classification_class=classification_class,
+                              classification_scope=classification_scope, clusters=clusters)
+        et = ElementTree.fromstring(xml)
+        device_type, problems = parse_single_device_type(et)
+        asserts.assert_equal(len(problems), 0, "Unexpected problems parsing device type conformance")
+        asserts.assert_equal(len(device_type.keys()), 1, "Unexpected number of device types returned")
+        asserts.assert_true(device_type_id in device_type.keys(), "device type id not found in returned data")
+        asserts.assert_equal(device_type[device_type_id].revision, revision, "Unexpected revision")
+        asserts.assert_equal(len(device_type[device_type_id].server_clusters), len(clusters), "Unexpected number of clusters")
+        for id, name in clusters.items():
+            asserts.assert_equal(device_type[device_type_id].server_clusters[id].name, name, "Incorrect cluster name")
+            asserts.assert_equal(str(device_type[device_type_id].server_clusters[id].conformance),
+                                 'M', 'Incorrect cluster conformance')
+
+        clusters = {}
+        xml = template.render(device_type_id=device_type_id, revision=revision, classification_class=classification_class,
+                              classification_scope=classification_scope, clusters=clusters)
+        et = ElementTree.fromstring(xml)
+        device_type, problems = parse_single_device_type(et)
+        asserts.assert_equal(len(problems), 0, "Unexpected problems parsing device type conformance")
+        asserts.assert_equal(len(device_type.keys()), 1, "Unexpected number of device types returned")
+        asserts.assert_true(device_type_id in device_type.keys(), "device type id not found in returned data")
+        asserts.assert_equal(device_type[device_type_id].revision, revision, "Unexpected revision")
+        asserts.assert_equal(len(device_type[device_type_id].server_clusters), len(clusters), "Unexpected number of clusters")
+
+        # Bad device type ID
```

Comment:
how about having each of these as individual tests rather than a single test. We could set up the template in the constructor or some setUp call.

Reply:
changed.